RFCs/FS-1063-support-letbang-andbang-for-applicative-functors.md

```diff
@@ -0,0 +1,432 @@
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The new applicative computation expressions are quite constrained, and as has been discussed, that is precisely what allows them to be so useful. However, these constraints are potentially somewhat unintuitive to the beginner. Computation expressions already involve one of the steeper learning curves of the F# language features, so the added complexity from this feature needs to be carefully weighed against the potential guarantees, expressiveness and performance gains that they can offer.
+
+# Alternative Designs
+[alternative-designs]: #alternative-designs
+
+[Tomas Petricek's Joinads](http://tomasp.net/blog/fsharp-variations-joinads.aspx/) offered a superset of the features proposed here, but was [rejected](https://github.com/fsharp/fslang-suggestions/issues/172) due to its complexity. This RFC is of much smaller scope, so should be a much less risky change.
+
+# Compatibility
+[compatibility]: #compatibility
+
+## Is this a breaking change?
+
+This change should be backwards compatible.
+
+Existing computation expression builders with an `Apply` method should not change in behaviour, since usages of the builder would still need to add the new `let! ... and! ...` syntax to activate it. In particular, in the case of `let! ... return ...`, we will continue to only pick bind, [as mentioned earlier](#singlelet).
+
+## What happens when previous versions of the F# compiler encounter this design addition as source code?
+
+Previous compiler versions reject the new `and!` keyword:
+
+```
+error FS1141: Identifiers followed by '!' are reserved for future use
+```
+
+## What happens when previous versions of the F# compiler encounter this design addition in compiled binaries?
+
+Since the syntax is desugared into a method call on the builder object, after compilation usages of this feature will be usable with previous compiler versions.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+Is `anduse!` an acceptable keyword for when `and!` must also imply a call to `Using`?
+
+There are various ways of desugaring `let! ... return ...` via one of `Map`, `Apply` or `Bind`. [The above](#singlelet) assumes that we are aiming for backward compatibility and hence doesn't consider desugaring to `Apply` at all. Is this the best choice? Alternatives include:
+
+* Using `Apply` in preference to `Bind` for `let! ... return` whenever `Apply` is defined (on the basis that that any reasonable `Apply` implementation will be functionally equivalent, but more efficient than the corresponding `Bind`)
+* Subsuming [the RFC for desugaring this to `Map`](https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1048-ce-builder-map.md) and defining a hierarchy between `Map`, `Apply` and `Bind` and some attributes for those methods to allow the creators of builders to opt-in to "optimisations" that pick the least powerful (and hence hopefully most efficient) desugaring.
```

Comment:
Another mention of powerful here that may need clarification. Perhaps doing so with a short glossary of terms of the top is better?